### DISCUSSION TITLE
typeof window !== undefined  => true  ????

### DISCUSSION THREAD

**Discussion by zcTob** [vercel/next.js#30556] [Help] [answered]
### What version of Next.js are you using?

latest

### What version of Node.js are you using?

v12.22.1

### What browser are you using?

chrome

### What operating system are you using?

macOs

### How are you deploying your application?

next start

### Describe the Bug

root add .babelrc 

code typeof window !== undefined  => true

### Expected Behavior

code `typeof window !== undefined`  => `typeof window !== undefined`

### To Reproduce

![image](https://user-images.githubusercontent.com/22114645/139279792-d12c08c3-33cf-4a3a-9756-d84e87f0224f.png)
![image](https://user-images.githubusercontent.com/22114645/139279814-ca5985a8-7ff5-4704-9926-c3c9c36aafa5.png)
![image](https://user-images.githubusercontent.com/22114645/139279862-5ab5615c-402f-4d7c-b620-acf55d026cf7.png)
![image](https://user-images.githubusercontent.com/22114645/139279879-dc2a5392-34e4-4154-99d0-d5b7e05e38ec.png)

**Comment by ijjk** [comment, marked as answer]
Hi, this is an optimization done to help with [dead-code elimination](https://en.wikipedia.org/wiki/Dead_code_elimination), we know during build-time that this will always be `true` for the client build and `false` for the server build so we replace it during the build.

**Reply by zcTob** [reply]
> Hi, this is an optimization done to help with [dead-code elimination](https://en.wikipedia.org/wiki/Dead_code_elimination), we know during build-time that this will always be `true` for the client build and `false` for the server build so we replace it during the build.

thanks

**Comment by zcTob** [comment]
@ijjk  How do I disable him

**Reply by ijjk** [reply]
Disable the automatic replacing? It can't be disabled currently as it could cause unexpected behavior

**Comment by zcTob** [comment]
i have a sharedworker , old code  has typeof window !== undefined,   He makes mistakes

**Comment by zcTob** [comment]
anything can help me？
shareworker.js has next code
hasWindow = （） = 》｛typeof window != undefined｝
hasWindow &amp;&amp; window.xxx
then devtool has error. &nbsp;window is not defined


thank for your help







发自我的iPhone


------------------ Original ------------------
From: JJ Kasper ***@***.***&gt;
Date: Thu,Oct 28,2021 11:11 PM
To: vercel/next.js ***@***.***&gt;
Cc: zyrepo ***@***.***&gt;, Author ***@***.***&gt;
Subject: Re: [vercel/next.js] typeof window !== undefined  =&gt; true  ???? (Discussion #30556)





 
Disable the automatic replacing? It can't be disabled currently as it could cause unexpected behavior
 
—
You are receiving this because you authored the thread.
Reply to this email directly, view it on GitHub, or unsubscribe.
Triage notifications on the go with GitHub Mobile for iOS or Android.

**Reply by ijjk** [reply]
Ah if the code is going inside of a worker thread `self` should be used instead of `window` and the `typeof window` check shouldn't be necessary
